include/aws/io/retry_strategy.h

```diff
@@ -0,0 +1,175 @@
+#ifndef AWS_IO_CLIENT_RETRY_STRATEGY_H
+#define AWS_IO_CLIENT_RETRY_STRATEGY_H
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/io/event_loop.h>
+
+#include <aws/common/atomics.h>
+
+struct aws_retry_strategy;
+struct aws_retry_token;
+
+/**
+ * Invoked upon the acquisition, or failure to acquire a retry token. This function will always be invoked if and only
+ * if aws_retry_strategy_acquire_retry_token() returns AWS_OP_SUCCESS. It will never be invoked synchronously from
+ * aws_retry_strategy_acquire_retry_token().
+ */
+typedef void(aws_retry_strategy_on_retry_token_acquired_fn)(
+    struct aws_retry_strategy *retry_strategy,
+    int error_code,
+    struct aws_retry_token *token,
+    void *user_data);
+
+/**
+ * Invoked after a successful call to aws_retry_strategy_schedule_retry(). This function will always be invoked if and
+ * only if aws_retry_strategy_schedule_retry() returns AWS_OP_SUCCESS. It will never be invoked synchronously from
+ * aws_retry_strategy_schedule_retry().
+ */
+typedef void(aws_retry_strategy_on_retry_ready_fn)(struct aws_retry_token *token, int error_code, void *user_data);
+
+enum aws_retry_error_type {
+    /** This is a connection level error such as a socket timeout, socket connect error, tls negotiation timeout etc...
+     * Typically these should never be applied for non-idempotent request types. */
+    AWS_RETRY_ERROR_TYPE_TRANSIENT,
```

Comment:
[clarification] Not sure what you mean by "Typically these should never be applied for non-idempotent request types"

Do you mean that TRANSIENT is the only type where you're not sure whether the request/response completed? So it's the only one that's not safe for retrying non-idempotent actions?

Reply:
yes.